lib/src/null_safety_analysis.dart

```diff
@@ -0,0 +1,178 @@
+  /// used as a dependency from different packages.
+  ///
+  /// Furthermore by awaiting the Future stored here, we avoid race-conditions
+  /// from downloading the same package-version into [_systemCache]
+  /// simultaneously when doing concurrent analyses.
+  final Map<PackageId, Future<NullSafetyCompliance>>
+      _packageInternallyGoodCache = {};
+  static final _firstVersionWithNullSafety = Version.parse('2.10.0');
+
+  NullSafetyAnalysis(SystemCache systemCache) : _systemCache = systemCache;
+
+  /// Returns true if package version [packageId] and all its non-dev
+  /// dependencies (transitively) have a language version >= 2.10, and no files
+  /// in lib/ of  these packages opt out to a pre-2.10 language version.
+  ///
+  /// This will do a full resolution of that package's import graph, and also
+  /// download the package and all dependencies into [cache].
+  ///
+  /// To avoid race conditions on downloading to the cache, only one instance
+  /// should be computing nullSafetyCompliance simultaneously with the same
+  /// cache.
+  ///
+  /// If [packageId] is a relative path dependency [containingPath] must be
+  /// provided with an absolute path to resolve it against.
+  Future<NullSafetyCompliance> nullSafetyCompliance(PackageId packageId,
+      {String containingPath}) async {
+    // A space in the name prevents clashes with other package names.
+    final rootName = '${packageId.name} importer';
+    final root = Package.inMemory(Pubspec(rootName,
+        fields: {
+          'dependencies': {
+            packageId.name: {
+              packageId.source.name: packageId.source is PathSource
+                  ? (packageId.description['relative']
+                      ? path.join(containingPath, packageId.description['path'])
+                      : packageId.description['path'])
+                  : packageId.description,
+              'version': packageId.version.toString(),
+            }
+          }
+        },
+        sources: _systemCache.sources));
+
+    SolveResult result;
+    try {
+      result = await resolveVersions(
+        SolveType.GET,
+        _systemCache,
+        root,
+      );
+    } on SolveFailure {
+      return NullSafetyCompliance.analysisFailed;
+    }
+
+    var allPackagesGood = true;
+    for (final dependencyId in result.packages) {
+      if (dependencyId.name == root.name) continue;
+
+      final packageInternallyGood =
+          await _packageInternallyGoodCache.putIfAbsent(dependencyId, () async {
+        final boundSource = dependencyId.source.bind(_systemCache);
+        final pubspec = await boundSource.describe(dependencyId);
+        final languageVersion = _languageVersion(pubspec);
+        if (languageVersion == null ||
+            languageVersion < _firstVersionWithNullSafety) {
+          return NullSafetyCompliance.notCompliant;
+        }
+
+        if (boundSource is CachedSource) {
+          /// TODO(sigurdm): Should we set metadata here?
```

Comment:
what kind of metadata?

Reply:
The `X-Pub-Reason` header set by `withDependencyType`.

Reply:
ah, hmm... I guess it would be nice to set this correctly. Given that we know if it's a dependency, dev_dependency or transitive dependency.

In most cases we'll probably have the response cached locally when we get here. I'm not sure I can imagine a case where we won't have that. But for consistency it would be nice to setup :)
(I'm okay leaving it as a todo if we just document what it is we mean here).

Reply:
I clarified the TODO.